src/array_api_stubs/_draft/array_object.py

```diff
@@ -343,9 +341,52 @@ def __dlpack__(
 
         Notes
         -----
+        Major DLPack versions represent ABI breaks, minor versions represent
+        ABI-compatible additions (e.g., new enum values for new data types or
+        device types).
+
+        The ``max_version`` keyword was introduced in v2023.12, and goes
+        together with the ``DLManagedTensorVersioned`` struct added in DLPack
+        1.0. This keyword may not be used by consumers for some time after
+        introduction. It is recommended to use this logic in the implementation
+        of ``__dlpack__``:
+
+        .. code:: python
+
+            if max_version is None:
+                # Keep and use the DLPack 0.X implementation
+                # Note: from March 2025 onwards (but ideally as late as
+                # possible), it's okay to raise BufferError here
+            else:
+                # We get to produce `DLManagedTensorVersioned` now
+                if max_version >= our_own_dlpack_version:
+                    # Consumer understands us, just return a Capsule with our max version
+                elif max_version[0] == our_own_dlpack_version[0]:
+                    # major versions match, we should still be fine here -
+                    # return our own max version
```

Comment:
Q: Should the producer raise a warning here? This is when the producer/consumer are ABI compatible, but the consumer may not understand some of the DLPack flags set by the producer (ex: ignoring readonly, though this was added in 1.0 and should not happen -- just an example).

This is slightly different from the case that the consumer fully receives/understands the flags (ex: `max_version == our_own_dlpack_version`), but for various design reasons just can't honor it (ex: CuPy/PyTorch will likely ignore readonly). In that case the consumer should raise a warning.

Reply:
I think the producer emitting a warning would be potentially annoying, and I don't see much of an upside given that the consumer can check the exact version inside the returned `DLManagedTensorVersioned`. If the consumer decides it's a good idea to warn, then it can do so - but I personally wouldn't.

Of course if there's a flag added that's semantically meaningful where it really would matter, then sure. But I can't think of concrete examples, most version mismatches will be support for new device types or dtypes.

Reply:
OK, resolving.